### PR TITLE
feat(gax): add Stream auto pagination

### PR DESCRIPTION
**Pull request by codyoss**
This addes a new Paginator type that can be exposed by List RPCs for auto paginating over async streams. This feature is not yet stable in rust so we should have users opt-in to using this unstable feature. When Streams are added to the stdlib we will likely want to make a breaking change to use the traits defined there.

This currently only auto-paginates over pages. We will need to expand the trait of PageableResponse to also describe items to iterate over to paginate over those. But this seemed like enough complexity for one PR.

Updates: #297

### CODE REVIEW

File: src/gax/src/paginator.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+impl<T, E> Paginator<T, E>
+where
+    T: PageableResponse,
+{
+    /// Creates a new [Paginator] given the initial page token and a function
+    /// to fetch the next [PageableResponse].
+    pub fn new<F>(seed_token: String, execute: impl Fn(String) -> F + Clone + 'static) -> Self
+    where
+        // TODO(codyoss): This, and other types, may also need send here in practice.
```
Comment (coryan) [inline src/gax/src/paginator.rs:42]: Please convert this to a GitHub issue. 

Can we test if this requires `Send` and/or `Sync`?

File: src/gax/src/lib.rs [hunk 2]
```diff
@@ -62,3 +62,8 @@ mod request_parameter;
 
 /// The core error types used by generated clients.
 pub mod error;
+
+/// Defines some types and traits to convert and use List RPCs as a Stream.
+/// Async streams are not yet stable, so neither is the use of this feature.
+#[cfg(feature = "stream")]
```
Comment (coryan) [inline src/gax/src/lib.rs:68]: Will this be a feature in every generated library then?

File: src/gax/src/paginator.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    }
+
+    /// Returns the next mutation of the wrapped stream.
+    pub fn next(&mut self) -> futures::stream::Next<'_, Self> {
+        StreamExt::next(self)
+    }
+}
+
+enum IterState {
```
Comment (coryan) [inline src/gax/src/paginator.rs:78]: This looks like a [std::ops::ControlFlow<String,()>](https://doc.rust-lang.org/beta/std/ops/enum.ControlFlow.html) Should we use that instead?

File: src/gax/src/paginator.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+{
+    /// Creates a new [Paginator] given the initial page token and a function
+    /// to fetch the next [PageableResponse].
+    pub fn new<F>(seed_token: String, execute: impl Fn(String) -> F + Clone + 'static) -> Self
+    where
+        // TODO(codyoss): This, and other types, may also need send here in practice.
+        F: Future<Output = Result<T, E>> + 'static,
+    {
+        let stream = unfold(IterState::Token(seed_token), move |state| {
+            let execute = execute.clone();
```
Comment (coryan) [inline src/gax/src/paginator.rs:46]: Why do we need to clone this? Feels like we should be able to borrow the function to call it, that is why it is not a `FnOnce`, right?

File: src/gax/src/paginator.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    struct InnerClient {
+        data: Arc<Mutex<Vec<TestResponse>>>,
+    }
+
+    impl Client {
+        fn build_outgoing_list_request(&self) {
+            // placeholder for building an http request
+        }
```
Comment (coryan) [inline src/gax/src/paginator.rs:132]: I am not sure this adds enough value for a test.

File: src/gax/src/paginator.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            },
+        ];
+
+        let client = Client {
+            inner: Arc::new(InnerClient {
+                data: Arc::new(Mutex::new(responses)),
+            }),
+        };
+        let mut resps = vec![];
+        let mut stream = client.list_rpc_stream(TestRequest::default());
```
Comment (coryan) [inline src/gax/src/paginator.rs:197]: Seems like you are missing a test where the initial token is not the empty string.

**Review by coryan — APPROVED** [review]
Some minor stuff.  I have questions, but they are really for my education, not intended to block you.

File: src/gax/src/paginator.rs (hunk 4 above)
Comment (codyoss) [reply src/gax/src/paginator.rs:47]: I believe this is needed because Fn does not implement copy and we move this this function  to the async block below for the call. There may be a better way to call this, but this is where I ended up after a well fought battle with the compiler. If you see away around this I am open, just could not figure out anything better.
